ARROW-15366: [Docs] Automate incrementing of package version for R and non-R version switchers

thisisnic opened the pull request.

jorisvandenbossche commented:
Can you print an example of what json file the scripts outputs eg for the 7.0 release?

thisisnic commented:
> Can you print an example of what json file the scripts outputs eg for the 7.0 release?

Certainly can.  For the main docs, you manually add the dev version in the JS, right?

```
[
    {
        "name": "7.0 (stable)",
        "version": ""
    },
    {
        "name": "6.0",
        "version": "6.0/"
    },
    {
        "name": "5.0",
        "version": "5.0/"
    },
    {
        "name": "4.0",
        "version": "4.0/"
    },
    {
        "name": "3.0",
        "version": "3.0/"
    },
    {
        "name": "2.0",
        "version": "2.0/"
    },
    {
        "name": "1.0",
        "version": "1.0/"
    }
]
```

jorisvandenbossche commented:
>  For the main docs, you manually add the dev version in the JS, right?

Ah, no, it would actually be good this is handled by the script as well. I added it manually last week directly in the hosted sources (https://github.com/apache/arrow-site/pull/177) to get the dropdown working for the new nightly docs that were added (originally when adding the dropdown, we didn't yet have dev docs). But I should actually have ported that back to the main version of the json file in the apache/arrow repo)

thisisnic commented:
> > For the main docs, you manually add the dev version in the JS, right?
> 
> Ah, no, it would actually be good this is handled by the script as well. I added it manually last week directly in the hosted sources ([apache/arrow-site#177](https://github.com/apache/arrow-site/pull/177)) to get the dropdown working for the new nightly docs that were added (originally when adding the dropdown, we didn't yet have dev docs). But I should actually have ported that back to the main version of the json file in the apache/arrow repo)

Just pushed a commit which now results in this; is this what you'd expect?

```
[
    {
        "name": "8.0 (dev)",
        "version": "dev"
    },
    {
        "name": "7.0 (stable)",
        "version": ""
    },
    {
        "name": "6.0",
        "version": "6.0/"
    },
    {
        "name": "5.0",
        "version": "5.0/"
    },
    {
        "name": "4.0",
        "version": "4.0/"
    },
    {
        "name": "3.0",
        "version": "3.0/"
    },
    {
        "name": "2.0",
        "version": "2.0/"
    },
    {
        "name": "1.0",
        "version": "1.0/"
    }
]
```

jorisvandenbossche commented:
Yes, that looks good!

thisisnic commented:
I'm a bit stuck on this as I updated all the hunks in the Ruby tests, but I have another error where an assertion fails (https://github.com/apache/arrow/runs/4957293229?check_suite_focus=true#step:6:25).  I've tried to run the scripts locally, but am getting archery errors, e.g. 

```
(base) nic2@xps:~/arrow$ dev/release/01-prepare.sh 7.0.0 8.0.0 15
Delete existing git tag apache-arrow-7.0.0
Deleted tag 'apache-arrow-7.0.0' (was be397071f)
Create local branch release-7.0.0-rc15 for release candidate 15
Switched to a new branch 'release-7.0.0-rc15'
Updating changelog for 7.0.0
Traceback (most recent call last):
  File "/home/nic2/.local/lib/python3.8/site-packages/git/util.py", line 889, in __getitem__
    return getattr(self, index)
  File "/home/nic2/.local/lib/python3.8/site-packages/git/util.py", line 882, in __getattr__
    return list.__getattribute__(self, attr)
AttributeError: 'IterableList' object has no attribute 'apache-arrow-6.0.0'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/nic2/anaconda3/bin/archery", line 33, in <module>
    sys.exit(load_entry_point('archery', 'console_scripts', 'archery')())
  File "/home/nic2/anaconda3/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/nic2/anaconda3/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/nic2/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/nic2/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/nic2/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/nic2/anaconda3/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/nic2/anaconda3/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/nic2/anaconda3/lib/python3.8/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/home/nic2/arrow/dev/archery/archery/cli.py", line 832, in release_changelog_add
    changelog = release.changelog()
  File "/home/nic2/arrow/dev/archery/archery/release.py", line 408, in changelog
    curation = self.curate()
  File "/home/nic2/arrow/dev/archery/archery/release.py", line 386, in curate
    for c in self.commits:
  File "/home/nic2/arrow/dev/archery/archery/release.py", line 365, in commits
    lower = self.repo.tags[self.previous.tag]
  File "/home/nic2/.local/lib/python3.8/site-packages/git/util.py", line 891, in __getitem__
    raise IndexError("No item found with id %r" % (self._prefix + index)) from e
IndexError: No item found with id 'apache-arrow-6.0.0'
```
I'm a bit stuck in terms of what to try next!

kou commented on dev/release/test-helper.rb:
```diff
@@ -78,13 +78,16 @@ def detect_versions
     @snapshot_version = cpp_cmake_lists.read[/ARROW_VERSION "(.+?)"/, 1]
     @snapshot_major_version = @snapshot_version.split(".")[0]
     @release_version = @snapshot_version.gsub(/-SNAPSHOT\z/, "")
+    @release_minor_version = @release_version.split(".")[0]+"."+@release_version.split(".")[1]
```
This is not a minor version. This is a "major + minor" version.
How about (backward) "compatible" version?

```ruby
@release_compatible_version = @release_version.split(".")[0, 2]).join(".")
```

kou commented on dev/release/test-helper.rb:
```diff
@@ -78,13 +78,16 @@ def detect_versions
     @so_version = compute_so_version(@release_version)
     @next_version = @release_version.gsub(/\A\d+/) {|major| major.succ}
     @next_major_version = @next_version.split(".")[0]
+    @next_minor_version = @next_major_version + "."+ @next_version.split(".")[1]
```
ditto.

kou commented on dev/release/test-helper.rb:
```diff
@@ -78,13 +78,16 @@ def detect_versions
     @next_so_version = compute_so_version(@next_version)
     r_description = top_dir + "r" + "DESCRIPTION"
     @previous_version = r_description.read[/^Version: (.+?)\.9000$/, 1]
+    @previous_minor_version = @previous_version.split(".")[0]+"."+@previous_version.split(".")[1]
```
ditto.

kou commented on dev/release/utils-prepare.sh:
```diff
@@ -155,4 +155,10 @@ update_versions() {
   popd
+
+  pushd "${ARROW_DIR}"
+  python "./ci/scripts/util_update_docs_versions.py" . "${version}" "${r_version}"
```
We should ensure using Python 3 not Python 2 because Python2's `json` puts a trailing space after comma: https://bugs.python.org/issue16333

```suggestion
  ${PYTHON:-python3} "./ci/scripts/util_update_docs_versions.py" . "${version}" "${r_version}"
```

kou commented on dev/release/utils-prepare.sh:
```diff
@@ -155,4 +155,10 @@ update_versions() {
+  python "./ci/scripts/util_update_docs_versions.py" . "${version}" "${r_version}"
+  git add r/pkgdown/assets/versions.json
+  git add docs/source/_static/versions.json
+  popd
```
I think that we should only update `versions.json`s only when `${type}` is `release`.

kou reviewed (commented):
@jorisvandenbossche Why do you prefer to no trailing `/` in path for `dev`? https://github.com/apache/arrow-site/blob/asf-site/docs/_static/versions.json#L4
I want to add trailing `/` to `dev` for consistency because all other paths such as `5.0/` have trailing `/`.

@thisisnic It seems that your working copy doesn't have `apache-arrow-6.0.0` tag. `git fetch --all  --tags` will resolve it.

kou commented on dev/release/01-prepare-test.rb:
```diff
@@ -190,6 +190,22 @@ def test_version_pre_tag
+            "+        \"name\": \"#{@release_minor_version} (stable)\",",
+            "+    {",
+            "+        \"name\": \"#{@previous_minor_version}\",",
+            "+        \"version\": \"#{@previous_minor_version + "/"}\"",
```
```suggestion
            "+        \"version\": \"#{@previous_minor_version}/\"",
```

kou commented on dev/release/01-prepare-test.rb:
```diff
@@ -225,6 +241,20 @@ def test_version_pre_tag
+            "+        \"name\": \"#{@release_version} (release)\",",
+            "+    {",
+            "+        \"name\": \"#{@previous_version}\",",
+            "+        \"version\": \"#{@previous_minor_version + "/"}\"",
```
```suggestion
            "+        \"version\": \"#{@previous_minor_version}/\"",
```

jorisvandenbossche commented:
> @jorisvandenbossche Why do you prefer to no trailing `/` in path for `dev`? https://github.com/apache/arrow-site/blob/asf-site/docs/_static/versions.json#L4
> I want to add trailing `/` to `dev` for consistency because all other paths such as `5.0/` have trailing `/`.

Ah, that's not a preference, but rather a mistake I think :) Because without it, the url will be mis-formatted, and the dropdown will not go to the correct sub-page if choosing the dev version in the dropdown (the dropdown code checks if the url exists, and if not fallbacks to the home page of that version).

thisisnic commented:
Thanks for the feedback and suggestions here @kou !

Now entirely stuck again on test failures - the CI is showing differences between the expected line in one of the `json` files and the line that is there.

I've tried removing it entirely in 1 PR (see CI failure here: https://github.com/apache/arrow/runs/5040927092?check_suite_focus=true#step:6:403) and then copying and pasting the annotated "missing" line into the tests in a subsequent PR (see CI failure here: https://github.com/apache/arrow/runs/5040977997?check_suite_focus=true#step:6:405) but it's still showing up as not matching.

I've copied and pasted the two lines and done a diff on them and there's no difference, so I don't know what a sensible next step is here.

kou reviewed (approved):
+1

I've fixed the test failures. I've also fixed `versions.json` generation when `"X.Y (dev)"` exists.

kou commented:
I've also add missing trailing "/" to dev version.

thisisnic commented:
Thanks for finishing off this PR @kou, I was getting pretty frustrated with it, so your help here is much appreciated! :D
